pravega/pravega-operator
# Issue 60: Data protection / High availability

## Change log description
The operator should ensure that pods are deployed in a reasonable way from a data protection and high availability perspective.

#### Pod protection
Kubernetes has the ability to safely [evict all of your pods from a node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) before an admin can perform maintenance on a node. The operator should prevent this from happening if the node evicted would cause disruption in a Pravega cluster. This can be achieved with [Pod Disruption Budgets](https://kubernetes.io/docs/tasks/run-application/configure-pdb/).

With these changes, the operator will create a PDB for each component type (controller, segmentstore, bookie) with the following PDB configuration.

- Controller: allow planned disruptions as long as there is at one Controller pod available.
- Segment Store: allow 1 planned disruption if the number of replicas is greater than 1, otherwise, no planned disruption is allowed.
- Bookkeeper: allow 1 planned disruption at a time. 
  - The minimum number of Bookies is enforced to 3. 
  - The quorum has been brought down to 2 ([ref](https://github.com/pravega/pravega/pull/3158)) This allows one Bookie disruption even in the minimum configuration.
  - `AutoRecovery` is set to `true` by default to automatically recover under-replicated ledger when there is a pod disruption (planned or unplanned).

#### Pod placement

To achieve higher performance and fault tolerance instances of the same type are recommended to be spread across different nodes. For example, assuming that we have a 3-node Kubernetes cluster and we are attempting to create 3 Bookkeeper instances, the first Bookie will be placed in any of the 3 nodes (e.g. _Node 1_), the second one will be preferably placed in a different node (e.g. _Node 2_ or _Node 3_), and the third bookie will be placed in the only node that does not contain a bookie. If a fourth Bookie was to be scheduled, it would be placed in any of the nodes, because all of them already have one Bookie. 

This has been implemented in this PR with [Pod Anti-Affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#inter-pod-affinity-and-anti-affinity-beta-feature) using the `preferredDuringSchedulingIgnoredDuringExecution` soft requirement because there's no functional requirement to place components of the same kind in different nodes.

## Purpose of the change
- Fix #60 

## How to test the change
Assuming you already have access to a Kubernetes environment (GKE, PKS...) with at least 3 nodes, follow the instructions in the README file to deploy ZooKeeper, the NFS server provisioner, and the Pravega Tier 2 PVC. Skip the deployment of the Pravega operator.

Check out this branch.
```
git checkout issue-60-data-protection-ha
```

Update the Pravega operator image to use an image built with the PR changes. Open the `deploy/operator.yaml` file and replace the container image from `pravega/pravega-operator:latest` to `adrianmo/pravega-operator:issue-60`. 

Deploy the operator.
```
kubectl create -f deploy/operator.yaml
```
Follow the instructions in the README file to deploy a Pravega cluster with 3 Bookies, 1 or more Controllers, and 3 Segment stores.

Run the following command to verify the correct placement of pods.

```
$ kubectl get po -o wide
NAME                                         READY     STATUS    RESTARTS   AGE       IP           NODE
...
pravega-bookie-0                             1/1       Running   0          35s       10.56.6.43   gke-adrian-cluster-1-default-pool-6e9f73a5-7348
pravega-bookie-1                             1/1       Running   0          35s       10.56.1.29   gke-adrian-cluster-1-default-pool-0616952c-kd4z
pravega-bookie-2                             1/1       Running   0          35s       10.56.9.42   gke-adrian-cluster-1-default-pool-d0ef8647-f01m
pravega-operator-7dbd5bb757-wzp7z            1/1       Running   0          1h        10.56.9.38   gke-adrian-cluster-1-default-pool-d0ef8647-f01m
pravega-pravega-controller-58b4dcf4f-69jpv   1/1       Running   0          35s       10.56.1.28   gke-adrian-cluster-1-default-pool-0616952c-kd4z
pravega-segmentstore-0                       1/1       Running   0          35s       10.56.3.21   gke-adrian-cluster-1-default-pool-0616952c-n948
pravega-segmentstore-1                       1/1       Running   0          35s       10.56.9.41   gke-adrian-cluster-1-default-pool-d0ef8647-f01m
pravega-segmentstore-2                       1/1       Running   0          35s       10.56.6.42   gke-adrian-cluster-1-default-pool-6e9f73a5-7348
...
```
Verify that pods of the same kind are scheduled to different nodes.

Check that pod disruption budgets are correctly set.

```
$ kubectl get pdb
NAME                         MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
pravega-bookie               N/A             1                 1                     6m
pravega-pravega-controller   1               N/A               0                     6m
pravega-segmentstore         N/A             1                 1                     6m
```
In this example, we allow one pod disruption for Segment Store and Bookkeeper, but none disruption for the Controller since there's only one Controller pod.

If we try to drain the node that contains the Controller instance, we should expect the drain command to abort.

```
$ kubectl drain gke-adrian-cluster-1-default-pool-0616 --ignore-daemonsets
952c-kd4z                                                                                                                                                       
node "gke-adrian-cluster-1-default-pool-0616952c-kd4z" cordoned                                                                                                 
error: unable to drain node "gke-adrian-cluster-1-default-pool-0616952c-kd4z", aborting command...
```

```
$ kubectl get nodes
NAME                                              STATUS                     ROLES     AGE       VERSION
...
gke-adrian-cluster-1-default-pool-0616952c-kd4z   Ready,SchedulingDisabled   <none>    4d        v1.9.7-gke.11
...
```
The node will continue to be part of the cluster, but it will not host any new pods. The controller instance will still be untouched and the Pravega admin will be required to manually increase the number of replicas and/or delete the affected pod to automatically be rescheduled to a different node.
